rftp/server.go

```diff
@@ -28,11 +44,134 @@ func (s *Server) Listen(host string) error {
 
 	for {
 		msg := make([]byte, 1024)
-		_, addr, err := conn.ReadFromUDP(msg)
+		n, addr, err := conn.ReadFromUDP(msg)
 		if err != nil {
 			return err
 		}
 
-		conn.WriteToUDP(msg, addr)
+		go s.handlePacket(conn, addr, n, msg)
+	}
+}
+
+func (s *Server) handlePacket(conn *net.UDPConn, addr *net.UDPAddr, length int, packet []byte) {
+	header := &MsgHeader{}
+	header.UnmarshalBinary(packet)
+
+	switch header.msgType {
+	case msgClientRequest:
+		cr := &ClientRequest{}
+		err := cr.UnmarshalBinary(packet[header.hdrLen:])
+		if err != nil {
+			// TODO: Drop connection
+		}
+		s.connMgr.accept(conn, addr, cr, s.SRC)
+
+	case msgClientAck:
+		ack := &Acknowledgement{}
+		ack.UnmarshalBinary(packet[header.hdrLen:])
+		s.connMgr.handle(addr, ack)
+
+	case msgClose:
+		s.connMgr.close(addr)
+
+	default:
+		// TODO: Drop connection
+	}
+}
+
+type connection struct {
+	ch   chan *Acknowledgement
+	sock io.Writer
+}
+
+type connManager struct {
+	mux   sync.Mutex
+	conns map[ipKey]map[port]*connection
```

Comment:
I'd might make more sense to combine ip and port to one key to avoid nested maps. Using a string `ip:port` might also increase the readability of debug print statements. Or am I missing some advantage of the nested-map approach?

Reply:
done